plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/Exception.hs

```diff
@@ -81,47 +98,53 @@ mtraverse makeClassyPrisms
     , ''EvaluationError
     ]
 
-instance internal ~ MachineError fun => AsMachineError (EvaluationError user internal) fun where
-    _MachineError = _InternalEvaluationError
-instance AsUnliftingError internal => AsUnliftingError (EvaluationError user internal) where
-    _UnliftingError = _InternalEvaluationError . _UnliftingError
+instance structural ~ MachineError fun =>
+        AsMachineError (EvaluationError operational structural) fun where
+    _MachineError = _StructuralEvaluationError
+instance AsUnliftingError structural =>
+        AsUnliftingError (EvaluationError operational structural) where
+    _UnliftingError = _StructuralEvaluationError . _UnliftingError
 instance AsUnliftingError (MachineError fun) where
     _UnliftingError = _UnliftingMachineError
-instance AsEvaluationFailure user => AsEvaluationFailure (EvaluationError user internal) where
-    _EvaluationFailure = _UserEvaluationError . _EvaluationFailure
+instance AsEvaluationFailure operational =>
+        AsEvaluationFailure (EvaluationError operational structural) where
+    _EvaluationFailure = _OperationalEvaluationError . _EvaluationFailure
 
-type EvaluationException user internal =
-    ErrorWithCause (EvaluationError user internal)
+type EvaluationException operational structural =
+    ErrorWithCause (EvaluationError operational structural)
 
-{- Note [Ignoring context in UserEvaluationError]
-The UserEvaluationError error has a term argument, but
+{- Note [Ignoring context in OperationalEvaluationError]
+The OperationalEvaluationError error has a term argument, but
```

Comment:
I get what is happening here, but what I don't understand is why we do all this effort to distinquish between operational vs structural errors at actual script validation, by using
`Either (ErrorWithCause structural term) (EvaluationResult a)` . 

From the user's point of view, validating a script would either EvaluationSuccess (validate) or EvaluationFailure (not validate). 

In what way does it matter if the non-validation was caused by operational or structural error?

Reply:
It's explained in the Haddocks:

```
On the chain both of these are just regular failures and we don't distinguish between them there:
if a script fails, it fails, it doesn't matter what the reason was. However in the tests it does
matter why the failure occurred: a structural error may indicate that the test was written
incorrectly while an operational error may be entirely expected.
```

Basically, it's an extra level of safety for us. `unsafeToEvaluationResult` is used in quite a lot of places and some tests fail if you remove the distinction between operational and structural errors, in the same way that some tests will start to fail if you get typing wrong, even if operationally nothing changes. Just ensures that any evaluation failures are the expected operational ones, not somebody messing up the test.

Reply:
If it is an artifact of Tests alone, then perhaps EvaluationResult should belong to a testlib.

Because taking into account other Plutus-based languages (Aiken,Plutarch,etc) there may not be a distinction between structural vs operational errors and instead they have just errors?

Reply:
> If it is an artifact of Tests alone, then perhaps EvaluationResult should belong to a testlib.

Maybe. I'm currently removing it from builtins where we have a much more appropriate `BuiltinResult`. I doubt I'm going to get motivated enough to see it through though.

> Because taking into account other Plutus-based languages (Aiken,Plutarch,etc) there may not be a distinction between structural vs operational errors and instead they have just errors?

Perhaps, but I don't really care, I just want us to have that distinction for the same reasons I want us to have a type checker instead of juggling raw `Data`.

Reply:
> I doubt I'm going to get motivated enough to see it through though.

I mean, removing `EvaluationResult` from all non-test code, not replace it with `BuiltinResult` in the builtins code -- that ship has sailed.